kratos/utilities/volume_inside_voxel_utility.cpp

```diff
@@ -0,0 +1,382 @@
+                                                    Nodes[i].Z() + 0.5*v_left[2] + 0.5*v_right[2]));
+                PointsArrayType points;
+                points.push_back(Max_left);
+                points.push_back(&Nodes[i]);
+                points.push_back(Max_right);
+                points.push_back(Max_c);
+                double max_volume = TetraVolume(points)/FaceArea;
+
+                if (Case == 1) {
+                    left = 0.5;
+                    right = std::min(0.5,MinDistanceToNode[i].first);
+                } else if (Case == 2) {
+                    left = std::min(0.5,MinDistanceToNode[(i+3)%4].second);
+                    right = 0.5;
+                } else if (Case == 3) {
+                    left = std::min(0.5,MinDistanceToNode[(i+3)%4].second);
+                    right = std::min(0.5,MinDistanceToNode[i].first);
+                }
+                NodePtrType Int_left(new Node<3>(1, Nodes[i].X() + left*v_left[0], Nodes[i].Y() + left*v_left[1], Nodes[i].Z() + left*v_left[2]));
+                NodePtrType Int_right(new Node<3>(2, Nodes[i].X() + right*v_right[0], Nodes[i].Y() + right*v_right[1], Nodes[i].Z() + right*v_right[2]));
+                NodePtrType c(new Node<3>(2, Nodes[i].X() + left*v_left[0] + right*v_right[0], 
+                                                    Nodes[i].Y() + left*v_left[1] + right*v_right[1], 
+                                                    Nodes[i].Z() + left*v_left[2] + right*v_right[2]));
+                points.clear();
+                points.push_back(Int_left);
+                points.push_back(&Nodes[i]);
+                points.push_back(Int_right);
+                points.push_back(c);
+
+                if (Case != 0) {
+                    PartialArea =  max_volume -factor*TetraVolume(points)/FaceArea;
+                } else {
+                    PartialArea = 0;
+                }
+            }
+            Area += PartialArea;
+        }
+        return Area;    
+    }
+
+    /***********************************************************************************
+     **********************************************************************************/
+
+    double VolumeInsideVoxelUtility::Distance(const NodeType& Point0, const array_1d<double,3>& Point1) {
+        const double lx = Point0.X() - Point1[0];
+        const double ly = Point0.Y() - Point1[1];
+        const double lz = Point0.Z() - Point1[2];
+
+        const double length = lx * lx + ly * ly + lz * lz;
+
+        return std::sqrt( length );
+    }
+
+    /***********************************************************************************
+     **********************************************************************************/
+
+    double VolumeInsideVoxelUtility::TetraVolume(const PointsArrayType& rPoints) {
+        GeometryPtrType pGeom =Kratos::make_shared<Quadrilateral3D4<NodeType>>(rPoints);
+        return pGeom->Volume();
+    }
```

Comment:
This is extremely expensive. Use the closed formula from the point coordinates.

Reply:
> This is extremely expensive. Use the closed formula from the point coordinates.

I'm not using a perfect formula but an approximation (the 4 points are approximated as two triangles). It is cheaper and will work in a perfect way for most of the use cases of the utility.